### PR TITLE
8269697: JNI_GetPrimitiveArrayCritical() should not accept object array

### PR DESCRIPTION
**Pull request by zhengyu123**
Open this PR to carry on the discussion started in jdk17 [https://github.com/openjdk/jdk17/pull/185](url)

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8269697](https://bugs.openjdk.java.net/browse/JDK-8269697): JNI_GetPrimitiveArrayCritical() should not accept object array

### Reviewers
 * [Kim Barrett](https://openjdk.java.net/census#kbarrett) (@kimbarrett - **Reviewer**) ⚠️ Review applies to d6842ff99dd1a7c7dfe4f973d9dbd99a80c656f1
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to d6842ff99dd1a7c7dfe4f973d9dbd99a80c656f1

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4653/head:pull/4653` \
`$ git checkout pull/4653`

Update a local copy of the PR: \
`$ git checkout pull/4653` \
`$ git pull https://git.openjdk.java.net/jdk pull/4653/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4653`

View PR using the GUI difftool: \
`$ git pr show -t 4653`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4653.diff">https://git.openjdk.java.net/jdk/pull/4653.diff</a>

</details>

### CODE REVIEW

**Review by kimbarrett — APPROVED** [review]
Thanks for adding the test.  This looks good.

**Comment by kimbarrett** [comment]
Following upo on the discussion from https://github.com/openjdk/jdk17/pull/185

GetPrimitiveArrayCritical returns a void* which the caller is then expected
to cast to the appropriate pointer to jXXX for read or write.

If it's an objarray, it's not clear what type should be used for that cast;
it's not jobject.  JNI doesn't provide an appropriate type.

In fact, you need to know whether UseCompressedOops is enabled or not to
even begin to access values. Let's say you have side-channel knowledge of
the value of UseCompressedOops. Then what?

Writing has a good chance of leading to crashes with most (maybe all?) of our
collectors.  Even if you could somehow obtain a well-formed value, many
collectors have some barrier protocol that's needed for writes.

Reading and doing anything non-trivial with a value also has a good chance
of leading to crashes with at least some GCs.

I don't think changing it to return a copy helps much with any of that.
There are a few corner cases that might be different, but not interestingly
so. And JNI_COMMIT mode might have very bad results for some at least some
collectors.

So I'm going to claim there's no significant compatibility issue with code
that was violating the spec, because I don't think such code could have been
doing anything interesting or useful anyway. About all you can get from the
contents is some raw bits that you can't do much with.

File: test/hotspot/jtreg/runtime/jni/TestPrimitiveArrayCriticalWithBadParam/TestPrimitiveArrayCriticalWithBadParam.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+/**
+ * @test TestPrimitiveArrayCriticalWithBadParam
+ * @summary -Xcheck:jni should catch wrong parameter passed to GetPrimitiveArrayCritical
+ * @library /test/lib
+ * @run main/othervm/native TestPrimitiveArrayCriticalWithBadParam
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/jni/TestPrimitiveArrayCriticalWithBadParam/TestPrimitiveArrayCriticalWithBadParam.java:29]: othervm is not needed, but you will need to explicitly set the library path for the launched VM.

**Review by dholmes-ora — APPROVED** [review]
Hi Zhengyu,

Please put the test under runtime/jni/checked.

The actual code change is fine by me. I will discuss the compatibility issue elsewhere.

Thanks,
David

**Comment by dholmes-ora** [comment]
A CSR request has been created and filled out. @zhengyu123 please mark it as Finalized. Thanks. Others can comment on the CSR request. I have already included a large portion of @kimbarrett 's comment above.

**Pull request closed without merging**